agent/agentcontainers/containers_dockercli.go

```diff
@@ -31,6 +33,172 @@ func NewDocker(execer agentexec.Execer) Lister {
 	}
 }
 
+// ContainerEnvInfoer is an implementation of agentssh.EnvInfoer that returns
+// information about a container.
+type ContainerEnvInfoer struct {
+	container string
+	user      *user.User
+	userShell string
+	env       []string
+}
+
+// EnvInfo returns information about the environment of a container.
+func EnvInfo(ctx context.Context, execer agentexec.Execer, container, containerUser string) (*ContainerEnvInfoer, error) {
+	var dei ContainerEnvInfoer
+	dei.container = container
+
+	var stdoutBuf, stderrBuf bytes.Buffer
+	if containerUser == "" {
+		// Get the "default" user of the container if no user is specified.
+		// TODO: handle different container runtimes.
+		cmd, args := WrapDockerExec(container, "")("whoami")
+		execCmd := execer.CommandContext(ctx, cmd, args...)
+		execCmd.Stdout = &stdoutBuf
+		execCmd.Stderr = &stderrBuf
+		if err := execCmd.Run(); err != nil {
+			return nil, xerrors.Errorf("get container user: run whoami: %w: stderr: %q", err, strings.TrimSpace(stderrBuf.String()))
+		}
+		out := strings.TrimSpace(stdoutBuf.String())
+		if len(out) == 0 {
+			return nil, xerrors.Errorf("get container user: run whoami: empty output: stderr: %q", strings.TrimSpace(stderrBuf.String()))
+		}
+		containerUser = out
+		stdoutBuf.Reset()
+		stderrBuf.Reset()
+	}
+	// Now that we know the username, get the required info from the container.
+	// We can't assume the presence of `getent` so we'll just have to sniff /etc/passwd.
+	cmd, args := WrapDockerExec(container, containerUser)("cat", "/etc/passwd")
+	execCmd := execer.CommandContext(ctx, cmd, args...)
+	execCmd.Stdout = &stdoutBuf
+	execCmd.Stderr = &stderrBuf
+	if err := execCmd.Run(); err != nil {
+		return nil, xerrors.Errorf("get container user: read /etc/passwd: %w stderr: %q", err, strings.TrimSpace(stderrBuf.String()))
+	}
+
+	scanner := bufio.NewScanner(&stdoutBuf)
+	var foundLine string
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		if len(line) == 0 {
+			continue
+		}
+		if !strings.HasPrefix(line, containerUser+":") {
+			continue
+		}
+		foundLine = line
```

Comment:
I'd kind of expect a break here for clarity, although I'm pretty sure this "can't go wrong ™️". Comment would be fine too as to why we keep going.

Reply:
Let's take a break!